db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/FromBaseTable.java

```diff
@@ -902,18 +1322,61 @@ public CostEstimate estimateCost(OptimizablePredicateList predList,
                                      RowOrdering rowOrdering) throws StandardException {
         CostEstimate finalCostEstimate;
         CostEstimate firstPassCostEstimate;
+
+        // Unioned index scans access path
+        AccessPath uisAccessPath = null;
+
+        // Unioned Index Scans currently not eligible for semijoin.
+        // Build and cost the Unioned Index Scans access path.
+        OptimizablePredicate uisPred =
+            (isOneRowResultSet() || getCompilerContext().getDisableUnionedIndexScans()) ? null :
+            getUsefulPredicateForUnionedIndexScan(predList);
+        if (uisPred != null) {
+            // Using the found DNF predicate of index accesses, build the UNION
+            // statements to apply each ORed conditional expression in a separate
+            // scan, and combine all of the scans together.  The UNIONs will eliminate
+            // duplicate ROWIDs.
+            uisAccessPath =
+                buildUnionedScans(uisPred, optimizer);
+
+            // If a UIS UNION tree was built, construct the ROWID join(s) to get
+            // the final rows, optimize the tree, and cost it.
+            if (uisAccessPath != null)
+                bindAndOptimizeUnionedIndexScansPath(uisAccessPath, optimizer);
+        }
+
+        // Cost the standard base table access path.
         finalCostEstimate = firstPassCostEstimate =
             estimateCostHelper(predList, cd, outerCost, optimizer, rowOrdering);
         AccessPath currentAccessPath = getCurrentAccessPath();
 
-        // Cost accessing this conglomerate using the index with
-        // the first index column not used by any useful predicates
-        // versus scanning all rows in the conglomerate.
-        // Choose the cheapest of the two access paths.
         LanguageConnectionContext lcc = getLanguageConnectionContext();
-        if (lcc.favorIndexPrefixIteration())
-            return finalCostEstimate;
-        if (currentAccessPath.getNumUnusedLeadingIndexFields() > 0) {
+        currentAccessPath.setUisPredicate(null);
+        currentAccessPath.setUisRowIdPredicate(null);
+        currentAccessPath.setUnionOfIndexes(null);
+        currentAccessPath.setUisRowIdJoinBackToBaseTableResultSet(null);
+
+        boolean hintedUISAccessPath = false;
+
+        // Cost accessing this conglomerate using a Unioned Index Scans
+        // access path vs. standard access path.
+        if (uisAccessPath != null &&
+            (getCompilerContext().getFavorUnionedIndexScans() ||
+            uisAccessPath.getCostEstimate().compare(firstPassCostEstimate) < 0)) {
+
+            hintedUISAccessPath = getCompilerContext().getFavorUnionedIndexScans();
+            if (hintedUISAccessPath)
+                reduceEstimatedCosts(uisAccessPath);
+            currentAccessPath.copy(uisAccessPath);
+            finalCostEstimate = firstPassCostEstimate = uisAccessPath.getCostEstimate();
+        }
```

Comment:
This is a very hacky process. If I understand it correctly, the UIS access path is compared only to current access path, meaning current cd + current join strategy. Suppose we are in cost estimation of (table cd + NLJ) and then UIS has lower cost. Then we directly do `currentAccessPath.copy(uisAccessPath)`, meaning when calling `nextDecoratedPermutation()`, it would be (UIS + SortMerge join). So we skip all combinations of table cd with other join strategies! What if, for example, (table cd + Merge join) is has better cost than (UIS + NLJ)? My understanding is we would miss that chance.

Since we make an `AccessPath` for UIS, why don't we just integrate this directly into the framework:
```
while(optimizer.nextJoinOrder()){
    while(optimizer.getNextDecoratedPermutation()){
        optimizer.costPermutation();
    }
 }
```
Basically, in enumerating all conglomerates of this table, we could add one more fake conglomerate indicating the UIS access path. Internally, it could be just table cd + UIS flag. Then we would have all combinations of UIS fake cd with all join strategies enumerated, fitting into the current framework of deciding better plan.

Reply:
There should be no skipping of combinations in the current approach.  The UIS access path and associated cost is recalculated for each new join strategy we consider.  In your example, when switching from NLJ + UIS to SortMerge join via nextDecoratedPermutation(), all of the UIS-related fields in the access path are set to null and we begin again fresh with a new comparison.  I did notice a bug, however, where those access path fields were being set to null in the wrong location, so I moved them up to the top of FromBaseTable.estimateCost:

        currentAccessPath.setUisPredicate(null);
        currentAccessPath.setUisRowIdPredicate(null);
        currentAccessPath.setUnionOfIndexes(null);
        currentAccessPath.setUisRowIdJoinBackToBaseTableResultSet(null);

When looking at the access path enumeration flow, this is not as clean as it could be.  Really we should be clearing these fields out as soon as we set to a new access path.  Therefore I will also initialize these fields in FromTable.nextAccessPath.  Let me know if you know of more locations where new access paths are considered, where we should also reset these fields.

On the pass through the access paths which looks at the base table conglomerate we cost each join strategy, both with and without UIS.  If UIS wins for any given join strategy, it is saved into the best access path if it is cheaper than the current best access path.  As the next join is considered, UIS fields are reset, and we again cost normal access path vs. UIS access path, and also vs. IndexPrefixIteratorMode.  On subsequent passes, when only index conglomerates are considered, there is no need to cost UIS access path again because the same set of index conglomerates and primary keys used by UIS would be chosen again and any best cost UIS path is already stored in the best access path.

The alternative you mentioned to used a fake conglomerate and drive the combinations from the outside would also work, and I can see the appeal if your preference is to drive the index combinations from the outside.  But it seems a little like we'd be trying to fit this access path information into a ConglomerateDescriptor structure, which describes something entirely different.  We might not have useful information for every field, whereas building only the access path creates no made-up structures.  It seems like this is just a matter of preference.  Maybe we can have a discussion about this to compare the two approaches.